lib/src/strings.dart

```diff
@@ -67,142 +54,220 @@ String _yamlEncodeDoubleQuoted(String string) {
-  String result;
-
-  final trimmedString = string.trimRight();
-  final removedPortion = string.substring(trimmedString.length);
-
-  if (removedPortion.contains('\n')) {
-    result = '>+\n${' ' * indentation}';
-  } else {
-    result = '>-\n${' ' * indentation}';
-  }
-
-  /// Duplicating the newline for folded strings preserves it in YAML.
-  /// Assumes the user did not try to account for windows documents by using
-  /// `\r\n` already
-  return result +
-      trimmedString.replaceAll('\n', lineEnding * 2 + ' ' * indentation) +
-      removedPortion;
+/// See https://yaml.org/spec/1.2.2/#8112-block-chomping-indicator
+String _getChompingIndicator(String string) {
+  /// By default, we apply an indent to the string after every new line.
+  ///
+  /// Apply the `keep (+)` chomping indicator for trailing whitespace to be
+  /// treated as content.
+  ///
+  /// [NOTE]: We only check for trailing whitespace rather than `\n `. This is
+  /// a coin-toss approach. If there is a new line after this, it will be kept.
+  /// If not, nothing happens.
+  if (string.endsWith(' ') || string.endsWith('\n')) return '+';
+
+  return '-';
 }
 
-/// Generates a YAML-safe literal string.
+/// Attempts to encode a [string] as a _YAML folded string_ and apply the
+/// appropriate _chomping indicator_.
 ///
-/// It is important that we ensure that [string] is free of unprintable
-/// characters by calling [_hasUnprintableCharacters] before invoking this
-/// function.
-String _tryYamlEncodeLiteral(
-    String string, int indentation, String lineEnding) {
-  final result = '|-\n$string';
-
-  /// Assumes the user did not try to account for windows documents by using
-  /// `\r\n` already
-  return result.replaceAll('\n', lineEnding + ' ' * indentation);
-}
-
-/// Returns [value] with the necessary formatting applied in a flow context
-/// if possible.
+/// Returns `null`, if the [string] cannot be encoded as a _YAML folded
+/// string_.
 ///
-/// If [value] is a [YamlScalar], we try to respect its [YamlScalar.style]
-/// parameter where possible. Certain cases make this impossible (e.g. a plain
-/// string scalar that starts with '>'), in which case we will produce [value]
-/// with default styling options.
-String _yamlEncodeFlowScalar(YamlNode value) {
-  if (value is YamlScalar) {
-    assertValidScalar(value.value);
-
-    final val = value.value;
-    if (val is String) {
-      if (_hasUnprintableCharacters(val) ||
-          value.style == ScalarStyle.DOUBLE_QUOTED) {
-        return _yamlEncodeDoubleQuoted(val);
-      }
+/// **Examples** of folded strings.
+/// ```yaml
+/// # With the "strip" chomping indicator
+/// key: >-
+///   my folded
+///   string
+///
+/// # With the "keep" chomping indicator
+/// key: >+
+///   my folded
+///   string
+/// ```
+///
+/// See: https://yaml.org/spec/1.2.2/#813-folded-style
+String? _tryYamlEncodeFolded(String string, int indentSize, String lineEnding) {
```

Comment:
Also I'm realizing that we might be missing the point with folded mode.

The point with folded mode is that we can replace `a b` with `a\nb` and still get the same value.

If the string contains newlines then this becomes complicated, because a newline essentially have to be replaced by an empty-line, right?

But also, rather importantly: the point of a folded string is to be able to wrap long lines without affecting the end-result.

```yaml
A: |-
  my very long string that cannot be broken, if I don't want newlines.
# As folded string
B: >-
  my very long string that
  cannot be broken, if I
  don't want newlines.
```

In this case (A) and (B) are the same. The reason for using folded strings is that you don't want VERY long lines.

So if we want to support folded string, I guess we ought to maybe wrap strings at 80 characters, or something like that?

This is perhaps something we can do in a follow up PR, if we just make the folded string logic work correctly, then perhaps improving the formatting is a follow up.